Add paths for built-in types

yallop opened the pull request:
It's often useful to write or generate code that refers to built-in types such as `int` or `option`, but it's currently inconvenient to do so in a way that avoids shadowing.

This PR adds aliases in `Stdlib` for the built-in types `option`, `int`, `bool`, `unit`, and `exn`, and aliases in `List` and `Array` for the built-in types `list` and `array`.  Other built-in types (`floatarray`, `bytes`, `string`, `float`, `nativeint`, `int32`, `int64`, `lazy_t` and `char`) are already aliased elsewhere in the standard library.

At some point in the future there may be standard library modules for some of these types (e.g. `Option` or `Bool`); if that happens before the next release then the corresponding aliases in `Stdlib` could be removed.

(Fixes [Mantis 5072: Absolute name for built-in types such as int, bool, ... ?](https://caml.inria.fr/mantis/view.php?id=5072) and [Mantis 6655: Alias built-in types and pre-defined exceptions in Pervasives](https://caml.inria.fr/mantis/view.php?id=6655).)

yallop committed:
Add aliases for the built-in 'list' and 'array' types.

yallop committed:
Changes entry: built-in type aliases

v-gb commented on stdlib/stdlib.mli:
```diff
@@ -1065,6 +1065,30 @@ module LargeFile :
+
+(** {1 Exceptions} *)
+
+type nonrec exn = exn
```
Shouldn't this be: `type nonrec exn = exn = ..`? As it stands, I think your change breaks this program:

```ocaml
module M : sig type t = .. end = struct type t = exn = .. end;;
```

yallop replied:
Yes; fixed.

v-gb commented on stdlib/list.mli:
```diff
@@ -26,6 +26,9 @@
    longer than about 10000 elements.
 *)
 
+type nonrec 'a t = 'a list = [] | (::) of 'a * 'a list
```
`ListLabels` and `ArrayLabels` should be updated accordingly.

yallop replied:
Done.

v-gb commented:
Up to the couple of things I pointed out, the change does almost as advertised, the only exception being that the type `extension_constructor` is not aliased anywhere. Might be worth adding in `Obj` for consistency?
No strong opinion, but it could fine to alias all built-in type constructors in stdlib.mli, so that reading stdlib.mli tells you the initial scope. For this reason, I see no issue with having Stdlib.option and possibly in the future, Stdlib.Option.t.

I'll note that although this enable uses of the type constructors Stdlib.bool, the value constructors cannot be used in some cases, which theoretically limits what can be done without triggering warning 40 and 42 (which are enabled by default, even if I always disable them):

```ocaml
# type bla = [] | (::) | () | None | Some | false | true;;
type bla = [] | (::) | () | None | Some | false | true
# Stdlib.[];;
- : bla = [] (* wrong: scope issue *)
# let v = Stdlib.(false);;
val v : bool = Stdlib.false (* good, except the syntax "Stdlib.false" doesn't parse *)
# if true then false else Stdlib.(false);;
- : bla = false (* wrong: type based disambiguation apparently overrides module path *)
# if true then false else v;; (* good *)
Error: This expression has type bool but an expression was expected of type
         bla
```

And talking about constructors, the built in constructors of sum types are now available in the stdlib, but not the ones of exceptions (Not_found, Match_failure, Out_of_memory etc). I think you'd need to do this to claim to fix MPR#6655.

v-gb commented on stdlib/stdlib.mli:
```diff
@@ -1065,6 +1065,30 @@ module LargeFile :
   regular integers (type [int]), these alternate functions allow
   operating on files whose sizes are greater than [max_int]. *)
 
```
Wouldn't it make more sense to put all these aliases much earlier in the signature, in particular before they are used?

gasche commented:
I'm not sure what you mean by "wrong" in some of the examples you list; do you mean that the current semantics is wrong, or that the behavior is too surprising and we should warn about it, or something else?

Note that

```
# Stdlib.[];;
- : bla = []
```

comes from the fact, I think, that Jeremy re-exports `[]` in `Stdlib.List`, not in `Stdlib` itself. Since https://github.com/ocaml/ocaml/commit/1c4ee3a41aa0ff920289bad816278a99c5acfe22, which was merged in 4.04,`M.[]` is resolved as `let open M in []` (where `[]` is handled as an identifier `"[]"` by the parser).

v-gb commented:
Oh well I expected []/true/false to be normal constructors. But you're saying they are not, because there is no syntax to access a prefixed version of them, and the syntax that looks like that is actually a local open, so that explains the weird scoping.
Alright, well nevermind this remark.

yallop committed:
Add an Extension_constructor submodule to Obj.

Deprecate top-level functions extension_constructor / extension_name /
extension_id.

yallop commented:
Thanks for the thorough review, @sliquister.  I've addressed your comments in the latest set of pushed commits.

A number of tests are currently failing due to changes in type printing.  I plan to wait a little while for further comments on the existing change before addressing the tests to avoid churn.

yallop commented:
> the type `extension_constructor` is not aliased anywhere. Might be worth adding in `Obj` for consistency

Done.



> And talking about constructors, the built in constructors of sum types are now available in the stdlib, but not the ones of exceptions (`Not_found`, `Match_failure`, `Out_of_memory` etc). I think you'd need to do this to claim to fix MPR#6655.

Done.

Octachron commented:
Another beneficial effect of this PR is that it solves the issue with the manual index [MPR#7708](https://caml.inria.fr/mantis/view.php?id=7708) by giving an indexed path to built-in exceptions and types.

gasche commented:
One thing I hadn't commented on is: I'm not super happy about the not-real-constructors `[]`, `()`, `true` and `false` showing up in actual OCaml code. It is an open secret that these parse, and some people that are too clever for their own good have used it in the wild already, but having this right in Stdlib may hide the fact that this is an undocumented, not-guaranteed-to-keep-working thing the parser currently does.

Could you maybe add a comment whenever this happen, to indicate that this is a hack to give a path in Stdlib to those constants, but similar tricks are not meant to be used in user code?

yallop commented:
@gasche
> I'm not super happy about the not-real-constructors `[]`, `()`, `true` and `false` showing up in actual OCaml code.
[...]
> Could you maybe add a comment whenever this happen

I'm happy to add a comment for `()`, `true` and `false`.  But `[]` is a real constructor [nowadays](https://github.com/ocaml/ocaml/pull/234), isn't it?

gasche commented:
Not in my book (but I would be wrong if it was, for example, documented in the language reference); I thought of this PR more as an exercise in making compromises on the implementation details that helps advanced users that do not-really-supported things.

yallop commented:
The manual [has](https://github.com/ocaml/ocaml/blob/76db83e38476a109550fdcd7ca07440b4df43955/manual/manual/refman/typedecl.etex#L49-L50):

```
constr-decl:
(constr-name || '[]' || '(::)') [ 'of' constr-args ]
```

Drup commented:
IMHO, if `[]` and `::` are not documented nowadays, then it's a documentation failure, not an undocumented feature. The fact that we can redefine them is extremely useful in many cases.

If you feel like this should stay undocumented, but still want to fulfill the use case properly, you are welcome to champion the first commit in my [patchset for infix constructors](https://github.com/ocaml/ocaml/pull/76). :)

gasche commented:
For the record, I am rather in favor of integrating `(::)` *and* your infix constructors in the language. My impression, however, was that there was no consensus on that, and that user-defined lists were more of a tolerated accident. I appear to be wrong, though, as `[]` and `(::)` are documented. Maybe I should ask around to know what people feel now.

Drup commented:
I would be happy to resurrect my patchset (or even a subset of it) if you feel the weather is more favorable.

v-gb commented on stdlib/stdlib.mli:
```diff
@@ -51,6 +58,68 @@ exception Exit
+(** Exception raised when none of the cases of a pattern-matching
+   apply. The arguments are the location of the match keyword in the
+   source code (file name, line number, column number). *)
+
```
Unfortunately, I suspect this is not quite enough: predef.ml adds these attributes [@ocaml.warn_on_literal_pattern] on exceptions, and given that we shadow the built-in ones, we probably break the warnings.

v-gb replied:
(and if that's correct, there's the same problem in the other stdlib, though in the .ml since it doesn't have a .mli)

alainfrisch replied:
Ping @yallop for the comment above.

yallop replied:
@sliquister is correct -- omitting the `[@ocaml.warn_on_literal_pattern]` attribute does break the warning -- so I've added the attribute to the exceptions with arguments.

v-gb commented on stdlib/stdlib.mli:
```diff
@@ -51,6 +58,68 @@ exception Exit
+   stack reaches its maximal size. This often indicates infinite or
+   excessively deep recursion in the user’s program. (Not fully
+   implemented by the native-code compiler.) *)
+
```
This description is a bit weird "exception raised by bytecode runtime. Not fully implemented in native compiler". The first sentence should probably be runtime-agnostic and then, the last sentence can say that this is best-effort in the native runtime.

v-gb replied:
Oh I realized you took the description from the manual. Well the manual is weird then, but it's probably not worth diverging.

alainfrisch replied:
Currently, the manual has a dedicated chapter for built-in types and predefined exceptions.  Does it make sense to keep it if they are all defined in the Stdlib?  Or perhaps just pointers to reexport definitions?  We should avoid text duplication is possible.

More general question: do we want to keep those predefined exceptions as part of the global scope even if Stdlib is not opened?  Maintainers of stdlib alternatives might want to hide, rename or put in different sub-modules some of those exceptions.    I guess that exceptions are predefined because they are required by the runtime system (and relying on OCaml code to register them is risky if the full stdlib is not linked); but perhaps one could have a syntax to refer to internal exceptions without "polluting" the global scope:

````ocaml
exception Stack_overflow = Stack_overflow[@@ocaml.builtin];;
````

Do you think it is worth the trouble?

v-gb reviewed (commented):
The code/documentation looks good, I think the only concern is whether the warning on `Failure "int_of_string"` and friends is broken.

damiendoligez commented:
About the weird constructors: it's probably time to resume working on [MPR#5936](https://caml.inria.fr/mantis/view.php?id=5936). Please have a look at the discussion there.

alainfrisch commented on stdlib/stdlib.ml:
```diff
@@ -68,15 +83,19 @@ let max x y = if x >= y then x else y
-(* Boolean operations *)
+(* Boolean type and operations *)
+
+type nonrec bool = bool = false | true
```
This should be moved (or copied) to Bool (so that one can write Bool.true).

alainfrisch commented on stdlib/stdlib.ml:
```diff
@@ -68,15 +83,19 @@ let max x y = if x >= y then x else y
-(* Integer operations *)
+(* Integer type and operations *)
+
+type nonrec int = int
```
We already have `Stdlib.Int.t`, so to be coherent with `Stdlib.Array.t`, etc, we might not want to re-expose directly `Stdlib.int` (one counterargument is that some values in `Stdlib` use `int`, but none use `array`).

alainfrisch commented on stdlib/stdlib.mli:
```diff
@@ -30,9 +30,16 @@ module Pervasives : sig
 
+(** {1 Option type} *)
+
+type nonrec 'a option = 'a option = None | Some of 'a
```
Same here, we already have Stdlib.Option.t.  Do we also want to add Stdlib.option?

yallop commented:
@alainfrisch: I've removed the `bool`, `int`, and `option` aliases from `Stdlib`, and added `true` and `false` to the `Bool` module.

alainfrisch commented on otherlibs/threads/stdlib.ml:
```diff
@@ -29,6 +27,7 @@ let () =
   register_named_value "Pervasives.array_bound_error"
     (Invalid_argument "index out of bounds")
 
+type nonrec exn = exn = ..
```
Out of curiosity, does it affect client code in any way if reexport is done as above rather than just `type nonrec exn = exn`?

yallop replied:
Without the `= ..` it's not possible to extend `Stdlib.exn` — i.e. the following would be an error:

```ocaml
type Stdlib.exn += E
```

alainfrisch replied:
Ok, thanks!  I was not aware that aliases were not expanded in this context.

alainfrisch commented:
@damiendoligez You expressed somewhere else your reluctance against adding more "toplevel" declarations to `Stdlib`.  Can we argue that the components added here (types and exceptions) are already part of the "global scope" anyway and thus assume you are ok with those additions?

yallop committed:
Add 'true' and 'false' to the definition of Bool.t

yallop commented:
> The code/documentation looks good, I think the only concern is whether the warning on `Failure "int_of_string"` and friends is broken.

This is now fixed, and I've rebased/squashed to clean up the history.  I've also opened #2133 to make the `@ocaml.warn_on_literal_pattern` attribute more meaningful for multi-argument constructors and non-trivial patterns.

~~I think this is now ready for merge, if CI passes.~~  A number of tests are failing, because of printing differences, like this:

> ```
>-        val print : unit -> unit
>+        val print : unit/1 -> Stdlib.unit
> ```

Suggestions on how best to address these failures are welcome.  Some possibilities:
1.  Update the tests to use `-short-paths`, so that `Stdlib.unit` is always replaced with `unit`
2. Add a `Unit` module and move the type alias there instead, so that the built-in type is not shadowed by default
3. Special-case `Stdlib` in type-printing code so that aliases are always resolved.

My preference is for `2`, but I'm happy to hear arguments for other approaches.

yallop committed:
Add an alias for 'exn' to Printexc.

yallop committed:
Add a Unit module.

yallop committed:
Add paths for built-in exceptions.

yallop commented:
Update: I moved `unit` and `exn` out of `Stdlib` into their own modules (`Unit` and `Printexc` respectively).

alainfrisch commented on stdlib/array.ml:
```diff
@@ -13,6 +13,9 @@
 (**************************************************************************)
 
+(* An alias for the type of arrays. *)
+type nonrec 'a t = 'a array
```
nonrec is useless here (and in other modules as well).

yallop replied:
Thanks.  I've removed the useless `nonrec` keywords.

alainfrisch commented on stdlib/printexc.mli:
```diff
@@ -15,6 +15,9 @@
 
 (** Facilities for printing exceptions and inspecting current call stack. *)
 
+type nonrec exn = exn = ..
```
Is the idea at some point to rename `Printexc` to `Exn` (and keep `Printexc` as a deprecated alias)?  If so, it would make sense to call the type `t`.

yallop replied:
That seems like a good long-term plan.  I've renamed the type `t`.

alainfrisch commented on stdlib/obj.mli:
```diff
@@ -86,6 +86,7 @@ val int_tag : int
 val out_of_heap_tag : int
 val unaligned_tag : int   (* should never happen @since 3.11.0 *)
 
+type nonrec extension_constructor = extension_constructor
```
What about creating an Extension_constructor sub-module, with a type `t` and three functions `of_val:'a -> t`, `name: t -> string` and `id: t -> int` (keeping the existing functions as deprecated aliases)?

yallop replied:
Good idea.  I've made that change.

alainfrisch commented on stdlib/obj.mli:
```diff
@@ -86,9 +86,19 @@ val int_tag : int
+  val [@inline always] id : t -> int
+end
 val extension_constructor : 'a -> extension_constructor
+  [@@ocaml.deprecated "Use Obj.Extension_constructor.of_val instead."]
```
According to #2121, we seem to be moving to this style:
````
"use Obj.Extension_constructor.of_val"
````

alainfrisch reviewed (approved):
LGTM.  Waiting for a few days before merging in case someone else wants to add something.

alainfrisch commented:
Thanks @yallop!

hhugo commented:
The module [Unit] needs to be exported in [stdlib.ml]
